rslib/src/decks/add.rs

```diff
@@ -1,17 +1,12 @@
 // Copyright: Ankitects Pty Ltd and contributors
 // License: GNU AGPL, version 3 or later; http://www.gnu.org/licenses/agpl.html
-use super::name::{immediate_parent_name, normalize_native_name};
+use super::name::immediate_parent_name;
 use crate::{error::FilteredDeckError, prelude::*};
-use std::borrow::Cow;
 
 impl Collection {
-    /// Normalize deck name and rename if not unique. Bumps mtime and usn if
+    /// Rename deck if not unique. Bumps mtime and usn if
     /// name was changed, but otherwise leaves it the same.
     pub(super) fn prepare_deck_for_update(&mut self, deck: &mut Deck, usn: Usn) -> Result<()> {
-        if let Cow::Owned(name) = normalize_native_name(&deck.name) {
-            deck.name = name;
-            deck.set_modified(usn);
-        }
         self.ensure_deck_name_unique(deck, usn)
```

Comment:
I believe we can drop the check for normalisation because it should already be ensure when a name is constructed.
Should this routine also go then and we call `ensure_deck_name_unique()` directly?

Reply:
I'm a bit wary of removing this, because it's covering not only the :: → x1f conversion, but also unicode normalization, and checks for blank components, and I'm a bit worried parts of the Rust code may accidentally cause those problems now or in the future. I'm not sure we can rely on the names coming from the DB to be valid for example - a previous Anki version or some external app may have written bad data into them, so a check on writing is probably safest.

Reply:
Since we already normalise when converting from human names, I would think that normalising in the other constructors would make the most sense. Now that you've hidden the string field, NativeDeckName can guarantee that it always contains a normalised machine name.
We only need to split from_native_str() into two different constructors: `from_root()` (normalises, especially stripping \x1f) and `from_native_str()` (normalises components, preserving hierarchy).
What do you think?

Reply:
We read decks a lot more than we write them - displaying the deck list requires reading them all in at once for example, and some users have thousands or even tens of thousands of them. If the from_native() constructor performed normalization then, we'd be doing extra work on every read operation, instead of only doing it when writing a change to the DB, so I'm not sure that would be the best approach?

Reply:
Ah, that makes sense. I'll add a new normalise method then.